### PR TITLE
Use signatures consistently in complete OpenQA::WebSockets::Client

### PR DESCRIPTION
**Pull request by okurz**

### CODE REVIEW

**Review by Martchus — APPROVED** [review]
Was it bothering you that you didn't have any contributions to that file so far? :-)

**Comment by okurz** [comment]
Exactly that :)

**Comment by okurz** [comment]
What does

> Can't use string ("OpenQA::WebSockets::Client") as a HASH ref while "strict refs" in use at /usr/lib/perl5/vendor_perl/5.26.1/Mojo/Base.pm line 145.

in https://app.circleci.com/pipelines/github/os-autoinst/openQA/7082/workflows/252d96fb-92ae-440e-9598-a689f5170a1a/jobs/66861/steps?invite=true#step-108-58 want to tell me?

File: lib/OpenQA/WebSockets/Client.pm [hunk 1]
```diff
@@ -29,29 +29,25 @@ my $IS_WS_SERVER_ITSELF;
 sub mark_current_process_as_websocket_server { $IS_WS_SERVER_ITSELF = 1; }
 sub is_current_process_the_websocket_server  { return $IS_WS_SERVER_ITSELF; }
 
-sub new {
-    my $class = shift;
+sub new ($class) {
```
Comment (perlpunk) [inline lib/OpenQA/WebSockets/Client.pm:32]:
```suggestion
sub new ($class, @args) {
```
and then later
```
$class->SUPER::new(@args);
```

Comment (okurz) [reply lib/OpenQA/WebSockets/Client.pm:32]: ok, cool. Could you explain the error message to me? Why does it say "Can't use string ("OpenQA::WebSockets::Client") as a HASH ref"?

Comment (perlpunk) [reply lib/OpenQA/WebSockets/Client.pm:32]: Before it was doing `my $class = shift`, so `"OpenQA::WebSockets::Client"` wasn't in `@_` anymore.
Now with the missing shift, it was passed as the second arg to `$class->SUPER::new(@_);`

Comment (okurz) [reply lib/OpenQA/WebSockets/Client.pm]: Ah, I see. Strange that even with `Carp::Always` I could not see any trace mention that line to be problematic

Comment (okurz) [reply lib/OpenQA/WebSockets/Client.pm]: anyway, fixed now thanks to your suggestion.

**Comment by perlpunk** [comment]
I guess it would make sense to change those subs too:
https://github.com/os-autoinst/openQA/pull/4052/files#diff-eb555fee69b775e261bcf4f470f7c94c6026b53a99e1a276adc6db2755808c31L24-L30

**Comment by okurz** [comment]
> I guess it would make sense to change those subs too:
> https://github.com/os-autoinst/openQA/pull/4052/files#diff-eb555fee69b775e261bcf4f470f7c94c6026b53a99e1a276adc6db2755808c31L24-L30

Ok, I assume you mean like with `sub()`.

What about

```perl
has client => sub { OpenQA::Client->new(api => shift->host // 'localhost') };
```
?

```perl
has client => sub($args) { OpenQA::Client->new(api => $args->host // 'localhost') };
```

yields `Too many arguments for subroutine 'OpenQA::WebSockets::Client::__ANON__' at /usr/lib/perl5/vendor_perl/5.26.1/Mojo/Base.pm line 78.` and I don't know how the function signature should be define for this method

**Comment by perlpunk** [comment]
I guess `sub ($self, @)` should be the right signature for accessors. @kraih ?

**Comment by okurz** [comment]
ok, sounds good. Applied that in https://github.com/os-autoinst/openQA/pull/4052/commits/48876866a4e843293d5373184889c4921c7c8162

File: lib/OpenQA/WebSockets/Client.pm [hunk 2]
```diff
@@ -21,37 +21,33 @@ use Carp 'croak';
 use OpenQA::Client;
 use OpenQA::Utils 'service_port';
 
-has host   => sub { $ENV{OPENQA_WEB_SOCKETS_HOST} };
-has client => sub { OpenQA::Client->new(api => shift->host // 'localhost') };
-has port   => sub { service_port('websocket') };
+has host   => sub ($self) { $ENV{OPENQA_WEB_SOCKETS_HOST} };
+has client => sub ($self, @) { OpenQA::Client->new(api => $self->host // 'localhost') };
```
Comment (kraih) [inline lib/OpenQA/WebSockets/Client.pm:25]: That's not an accessor, it's just a closure that lazily creates the initial value for the attribute. There should be no additional arguments getting passed to it.

**Comment by kraih** [comment]
[This](https://github.com/mojolicious/mojo/blob/d2ea3915ea4e1c23ed32f4f84637a2a52298bb14/lib/Mojo/Base.pm#L78) is the line that runs the closure, and there should be no possibility of it receiving more than one argument. Maybe you had a typo somewhere?

**Comment by okurz** [comment]
Thanks. No typo. Works now if I put `($self)` on all

**Review by perlpunk — APPROVED** [review]

File: lib/OpenQA/WebSockets/Client.pm [hunk 3]
```diff
@@ -21,37 +21,33 @@ use Carp 'croak';
...
-has port   => sub { service_port('websocket') };
+has host   => sub ($self) { $ENV{OPENQA_WEB_SOCKETS_HOST} };
+has client => sub ($self) { OpenQA::Client->new(api => $self->host // 'localhost') };
+has port   => sub ($self) { service_port('websocket') };
 
 my $IS_WS_SERVER_ITSELF;
-sub mark_current_process_as_websocket_server { $IS_WS_SERVER_ITSELF = 1; }
-sub is_current_process_the_websocket_server  { return $IS_WS_SERVER_ITSELF; }
+sub mark_current_process_as_websocket_server() { $IS_WS_SERVER_ITSELF = 1; }
+sub is_current_process_the_websocket_server()  { return $IS_WS_SERVER_ITSELF; }
```
Comment (kraih) [inline lib/OpenQA/WebSockets/Client.pm:30]: Odd that perltidy won't insert a whitespace here between the name and the signature.

**Review by kraih — APPROVED** [review]

**Review by Martchus — APPROVED** [review]

**Review by Martchus — APPROVED** [review]
Considering @kraih had a look at it and didn't object using signatures for attribute `sub`s I suppose it is ok.

**Comment by okurz** [comment]
fails in

```
RETRY=0 timeout -s SIGINT -k 5 -v $((9 * 2 * (0 + 1) ))m tools/retry prove -l --harness TAP::Harness::JUnit --timer t/full-stack.t
[20:37:49] t/full-stack.t .. 18/? Bailout called.  Further testing stopped:  clickElement: element not interactable at /home/squamata/project/t/lib/OpenQA/SeleniumTest.pm:81
```

**Comment by kalikiana** [comment]
@okurz Do you want to force-push again?

**Comment by okurz** [comment]
that won't help without fixing

```
Too few arguments for subroutine 'OpenQA::WebSockets::Client::send_msg'
```

from https://app.circleci.com/pipelines/github/os-autoinst/openQA/8351/workflows/b2f92315-69d5-4ee6-bf90-e6bdbafc7069/jobs/78987?invite=true#step-111-64

File: lib/OpenQA/WebSockets/Client.pm (hunk 3 above)
Comment (okurz) [reply lib/OpenQA/WebSockets/Client.pm]: I added an extra style check which ensures this, same as done for os-autoinst. Also I adjusted the style of existing and these occurences accordingly.

**Pull request merged**